File: array_sum.py
#!/usr/bin/env python

class ArraySum:
  def __init__(self, input):
    self.input = sorted(input, reverse=True)
    self.count = 0
    self.length = len(self.input)
    self.somas = []
    self.output = [0 for i in self.input]
    input = sorted(input)
    soma = 0
    for i in range(0, self.length):
      soma += input[i]
      self.somas.append(soma)
    self.somas = self.somas[::-1]

  def find(self, total, i = 0):
    self.count += 1

    if i >= self.length: return False
    if total > self.somas[i]: return False

    current = self.input[i]

    if total == current:
     self.output[i] = total
     return True

    if self.find(total, i+1):
      return True

    self.output[i] = current

    if self.find(total-current, i+1):
      return True

    self.output[i] = 0

  def result(self):
    return [x for x in self.output if x > 0]

import sys, time

sys.setrecursionlimit(5000)

input = [int(x) for x in sys.argv[2::]]
expected = int(sys.argv[1])

arraySum = ArraySum(input)

before = time.clock()
arraySum.find(expected)
print "%f" % (time.clock() - before)
print ' '.join([str(x) for x in arraySum.result()])

File: array_sum.rb
#!/usr/bin/env ruby

class ArraySum
  def initialize(input)
    @input = input.sort.reverse
    @count = 0
    @length = input.length
    @somas = []
    @output = [0] * @length
    soma = 0
    @input.reverse.each do |n|
      soma += n
      @somas.unshift(soma)
    end
  end

  def find(total, i = 0)
    @count += 1
    return false if i >= @length
    return false if total > @somas[i]
    current = @input[i]
    return !! @output[i] = total if total == current
    return true if find(total, i + 1)
    @output[i] = current
    return true if find(total - current, i + 1)
    @output[i] = 0

    return false
  end

  def result
    @output.select {|n| n > 0}
  end
end

module ArraySumExtension
  def array_sum(total)
    ar = ArraySum.new(self)
    ar.find(total)
    ar.result
  end
end

class Array
  include ArraySumExtension
end

require 'benchmark'

numeros = ARGV.map &:to_i
expected = numeros.shift

result = []
time = Benchmark.realtime { result = numeros.array_sum(expected) }
puts "%f" % time
puts result.join ' '
